cassandane/data/directory.ldif

```diff
@@ -21,6 +21,20 @@ dc: internal
 memberof: cn=group co,ou=groups,o=cyrus
 memberof: cn=group o,ou=groups,o=cyrus
 
+dn: uid=group c,dc=internal
+objectclass: user
+uid: group c
+dc: internal
+memberof: cn=cassandane,ou=groups,o=cyrus
+memberof: cn=otheruser,ou=groups,o=cyrus
+
+dn: uid=group co,dc=internal
+objectclass: user
+uid: group co
+dc: internal
+memberof: cn=cassandane,ou=groups,o=cyrus
+memberof: cn=otheruser,ou=groups,o=cyrus
+
```

Comment:
Hmm.  This is actually really weird from an LDAP point of view.  In a real LDAP system, groups are real, you wouldn't fake it like this.

If you have the directory schema this file uses (where group membership is a property of the user object, which is what Cyrus's LDAP implementation seems to assume), then you would index your user objectclass on memberof, and expand group memberships by searching on `(&(objectclass=user)(memberof=[the group you're interested in]))` and collect the user dn's from that.

On the other hand, if you have first class group objects (the schema I'm used to from prior job), you would have a group objectclass with a member attribute that contains its members' dns, and your group expansions are `(&(objectclass=group)(dn=[the group you're interested in]))` and you collect the user dn's from the returned member values.

The data is also wrong.  The original groups-by-user data says "cassandane" is a member of "c" and "co", "otheruser" is a member of "o" and "co".  You're missing "group o" altogether, and have "otheruser" in "group c", where it doesn't belong.  The point of this test data is that both users should be in multiple groups, but not in _all_ groups.  The groups have those bits in their names as a redundancy for detecting whether a weird result is from bad test data or bad code.

With this bad data, if the tests are passing, then either:
* LDAP PTS doesn't use this data, in which case it's useless, or,
* the LDAP tests don't exercise the new users-by-group code, or
* the tests don't adequately verify that groups-by-user and users-by-group results are consistent

The HTTPPTS tests are based on the LDAP ones.  If the group memberships there are different, that was a mistake when they were converted... turns out "c" and "o" are visually very similar, easy to misread 😓

Reply:
I changed all the LDAP code to make more sense in standard LDAP terms.  The ldap pts module really is very awful code :(  I removed some of the most egregious mis-holding of APIs but it could use some more cleanup.

I'm also not entirely sure if what I'm doing for group membership is "right" but it works fine for test purposes.  We could change to using a filtered scan instead.

Reply:
And both modules now test for this case :)

Reply:
I believe that the ldap pts module was written by the OpenLDAP folks.

Reply:
yep; it's more whether what I've done for the test case is idiomatic ldap group with username members.  Yeah, OK it looks like this:

https://www.linkedin.com/advice/0/what-best-practices-managing-ldap-groups-roles

```

dn: cn=Admins,ou=Groups,dc=example,dc=com

changetype: modify

add: member

member: uid=charlie,ou=Users,dc=example,dc=com
```

So I want to add 'member' as the field, and extract a "uid" component from it for each user

Reply:
See, this makes sense, I don't know how you got to the "hasmember" stuff you implemented... though this comment was an hour ago, the "hasmember" stuff might have been before that